File: 백준/Gold/25173. 용감한 아리의 동굴 대탈출/용감한 아리의 동굴 대탈출.py
from collections import deque

#input 받기처리
n,m=map(int,input().split())
g=[list(map(int,input().split())) for _ in range(n)]
hero_hp,hero_attack,boss_hp,boss_attack=map(int,input().split())
stones=set()
for x in range(n):
    for y in range(m):
        if g[x][y]==2:
            hero_x,hero_y=x,y
            g[x][y]=0
        elif g[x][y]==3:
            boss_x,boss_y=x,y
            g[x][y]=0
        elif g[x][y]==1:
            stones.add((x,y))

#동남서북
dir=[(0,1),(1,0),(0,-1),(-1,0)]

#초기 방향 설정
for i in range(4):
    dx,dy=dir[i]
    if hero_x==boss_x+dx and hero_y==boss_y+dy:
        hero_dir=boss_dir=i
        break

while True:
    #아리의 공격 시작
    boss_hp-=hero_attack
    if boss_hp<=0:
        print("VICTORY!")
        exit()
    
    #아리의 이동
    #아리가 움직이지 못할 경우를 대비해서, 보스의 현재 위치를 저장함
    hero_saved_x,hero_saved_y=boss_x,boss_y
    for k in range(4):
        dx,dy=dir[hero_dir]
        nx,ny=hero_x+dx,hero_y+dy
        if 0<=nx<n and 0<=ny<m and g[nx][ny]==0 and (nx,ny)!=(boss_x,boss_y):
            hero_saved_x,hero_saved_y=hero_x,hero_y
            hero_x,hero_y=nx,ny
            break
        else:
            hero_hp-=1
            hero_dir=(hero_dir+1)%4
            if hero_hp<=0:
                print("CAVELIFE...")
                exit()
    
    #보스의 공격
    temp_x,temp_y,temp_dir=boss_x,boss_y,boss_dir
    cnt=1
    if stones:
        try:
            while True:
                cnt+=1
                dx,dy=dir[temp_dir]
                for i in range(cnt//2):
                    temp_x+=dx
                    temp_y+=dy
                    if (temp_x,temp_y) in stones:
                        raise
                temp_dir=(temp_dir+1)%4
        except:
            q=deque([(temp_x,temp_y,boss_attack)])
            v=[[0 for _ in range(m)] for _ in range(n)]
            while q:
                x,y,attack=q.popleft()
                if v[x][y] or attack<=0:continue
                v[x][y]=attack
                if (x,y)==(hero_x,hero_y):
                    hero_hp-=attack
                    if hero_hp<=0:
                        print("CAVELIFE...")
                        exit()
                    break
                for dx,dy in dir:
                    nx,ny=x+dx,y+dy
                    if 0<=nx<n and 0<=ny<m and g[nx][ny]!=1 and (nx,ny)!=(boss_x,boss_y):
                        q.append((nx,ny,attack-1))
    
    #보스의 이동
    boss_x,boss_y,boss_dir=hero_saved_x,hero_saved_y,hero_dir

File: 백준/Gold/25173. 용감한 아리의 동굴 대탈출/용감한 아리의 동굴 대탈출.java
import java.io.*;
import java.util.*;

public class Main {
    static int n, m;
    static int[][] g;
    static int hero_hp, hero_attack, boss_hp, boss_attack;
    static int hero_x, hero_y, boss_x, boss_y;
    static int hero_dir, boss_dir;
    static Set<String> stones = new HashSet<>();

    static int[][] dir = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}}; // 동남서북

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        g = new int[n][m];

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                g[i][j] = Integer.parseInt(st.nextToken());
                if (g[i][j] == 2) {
                    hero_x = i;
                    hero_y = j;
                    g[i][j] = 0;
                } else if (g[i][j] == 3) {
                    boss_x = i;
                    boss_y = j;
                    g[i][j] = 0;
                } else if (g[i][j] == 1) {
                    stones.add(i + "," + j);
                }
            }
        }

        st = new StringTokenizer(br.readLine());
        hero_hp = Integer.parseInt(st.nextToken());
        hero_attack = Integer.parseInt(st.nextToken());
        boss_hp = Integer.parseInt(st.nextToken());
        boss_attack = Integer.parseInt(st.nextToken());

        setInitialDirection();

        while (true) {
            // 아리의 공격 시작
            boss_hp -= hero_attack;
            if (boss_hp <= 0) {
                System.out.println("VICTORY!");
                return;
            }

            // 아리의 이동
            int hero_saved_x = boss_x;
            int hero_saved_y = boss_y;
            for (int k = 0; k < 4; k++) {
                int nx = hero_x + dir[hero_dir][0];
                int ny = hero_y + dir[hero_dir][1];
                if (isValidMove(nx, ny) && !(nx == boss_x && ny == boss_y)) {
                    hero_saved_x = hero_x;
                    hero_saved_y = hero_y;
                    hero_x = nx;
                    hero_y = ny;
                    break;
                } else {
                    hero_hp--;
                    hero_dir = (hero_dir + 1) % 4;
                    if (hero_hp <= 0) {
                        System.out.println("CAVELIFE...");
                        return;
                    }
                }
            }

            // 보스의 공격
            if (!stones.isEmpty()) {
                if (bossAttack()) {
                    System.out.println("CAVELIFE...");
                    return;
                }
            }

            // 보스의 이동
            boss_x = hero_saved_x;
            boss_y = hero_saved_y;
            boss_dir = hero_dir;
        }
    }

    static void setInitialDirection() {
        for (int i = 0; i < 4; i++) {
            int dx = dir[i][0], dy = dir[i][1];
            if (hero_x == boss_x + dx && hero_y == boss_y + dy) {
                hero_dir = boss_dir = i;
                break;
            }
        }
    }

    static boolean isValidMove(int x, int y) {
        return x >= 0 && x < n && y >= 0 && y < m && g[x][y] == 0;
    }

    static boolean bossAttack() {
        int temp_x = boss_x, temp_y = boss_y, temp_dir = boss_dir;
        int cnt = 1;

        while (true) {
            cnt++;
            int dx = dir[temp_dir][0], dy = dir[temp_dir][1];
            for (int i = 0; i < cnt / 2; i++) {
                temp_x += dx;
                temp_y += dy;
                if (stones.contains(temp_x + "," + temp_y)) {
                    return bossAttackFromStone(temp_x, temp_y);
                }
            }
            temp_dir = (temp_dir + 1) % 4;
            if (cnt >= n * m) break;
        }
        return false;
    }

    static boolean bossAttackFromStone(int stone_x, int stone_y) {
        Queue<int[]> q = new LinkedList<>();
        boolean[][] visited = new boolean[n][m];
        q.add(new int[]{stone_x, stone_y, boss_attack});
        visited[stone_x][stone_y] = true;

        while (!q.isEmpty()) {
            int[] pos = q.poll();
            int x = pos[0], y = pos[1], attack = pos[2];

            if (x == hero_x && y == hero_y) {
                hero_hp -= attack;
                if (hero_hp <= 0) return true;
                break;
            }

            if (attack <= 0) continue;

            for (int[] d : dir) {
                int nx = x + d[0];
                int ny = y + d[1];
                if (nx >= 0 && nx < n && ny >= 0 && ny < m && g[nx][ny] != 1 && !visited[nx][ny] && !(nx == boss_x && ny == boss_y)) {
                    visited[nx][ny] = true;
                    q.add(new int[]{nx, ny, attack - 1});
                }
            }
        }
        return false;
    }
}